amaru/src/consensus/header_validation.rs

```diff
@@ -0,0 +1,179 @@
+use crate::{
+    consensus::{Point, ValidateHeaderEvent},
+    sync::PullEvent,
+};
+use gasket::framework::*;
+use miette::miette;
+use ouroboros::{ledger::LedgerState, validator::Validator};
+use ouroboros_praos::consensus::BlockValidator;
+use pallas_crypto::hash::Hash;
+use pallas_math::math::{FixedDecimal, FixedPrecision};
+use pallas_network::facades::PeerClient;
+use pallas_primitives::conway::Epoch;
+use pallas_traverse::MultiEraHeader;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use tracing::{info, trace};
+
+pub type UpstreamPort = gasket::messaging::InputPort<PullEvent>;
+pub type DownstreamPort = gasket::messaging::OutputPort<ValidateHeaderEvent>;
+
```

Comment:
// TODO: Design and implement a proper abstraction for slot arithmetic. See https://github.com/pragma-org/amaru/pull/26/files#r1807394364